File: db2-sidecar/CommandHandler.cs
// db2-sidecar/CommandHandler.cs
using System.Diagnostics;
using IBM.Data.DB2.Core;

namespace Db2Sidecar;

public static class CommandHandler
{
    public static async Task<Response> Handle(Request req, ConnectionManager cm)
    {
        return req.Cmd switch
        {
            "connect"          => await Connect(req, cm),
            "disconnect"       => Disconnect(req, cm),
            "ping"             => await Ping(req, cm),
            "execute"          => await Execute(req, cm),
            "list_schemas"     => await ListSchemas(req, cm),
            "list_tables"      => await ListTables(req, cm),
            "get_table_schema" => await GetTableSchema(req, cm),
            _                  => new Response { Id = req.Id, Ok = false, Error = $"unknown_cmd:{req.Cmd}" }
        };
    }

    private static async Task<Response> Connect(Request req, ConnectionManager cm)
    {
        var err = await cm.Connect(req.ConnId!, req.ConnString!, req.Username!, req.Password!);
        return err is null
            ? new Response { Id = req.Id, Ok = true }
            : new Response { Id = req.Id, Ok = false, Error = err };
    }

    private static Response Disconnect(Request req, ConnectionManager cm)
    {
        cm.Disconnect(req.ConnId!);
        return new Response { Id = req.Id, Ok = true };
    }

    private static async Task<Response> Ping(Request req, ConnectionManager cm)
    {
        var conn = cm.Get(req.ConnId!);
        if (conn is null)
            return new Response { Id = req.Id, Ok = false, Error = "conn_not_found" };
        try
        {
            using var cmd = new DB2Command("SELECT 1 FROM SYSIBM.SYSDUMMY1", conn);
            await cmd.ExecuteScalarAsync();
            return new Response { Id = req.Id, Ok = true };
        }
        catch (Exception ex)
        {
            return new Response { Id = req.Id, Ok = false, Error = ex.Message };
        }
    }

    private static async Task<Response> Execute(Request req, ConnectionManager cm)
    {
        var conn = cm.Get(req.ConnId!);
        if (conn is null)
            return new Response { Id = req.Id, Ok = false, Error = "conn_not_found" };
        return await RunSql(req.Id, conn, req.Sql!);
    }

    private static async Task<Response> ListSchemas(Request req, ConnectionManager cm)
    {
        var conn = cm.Get(req.ConnId!);
        if (conn is null)
            return new Response { Id = req.Id, Ok = false, Error = "conn_not_found" };
        const string sql =
            "SELECT DISTINCT SCHEMANAME FROM SYSCAT.SCHEMATA " +
            "WHERE DEFINERTYPE = 'U' ORDER BY SCHEMANAME";
        return await RunSql(req.Id, conn, sql);
    }

    private static async Task<Response> ListTables(Request req, ConnectionManager cm)
    {
        var conn = cm.Get(req.ConnId!);
        if (conn is null)
            return new Response { Id = req.Id, Ok = false, Error = "conn_not_found" };
        var schema = req.Schema!.Replace("'", "''");
        var sql = $"SELECT TABNAME, TYPE FROM SYSCAT.TABLES WHERE TABSCHEMA = '{schema}' ORDER BY TABNAME";
        return await RunSql(req.Id, conn, sql);
    }

    private static async Task<Response> GetTableSchema(Request req, ConnectionManager cm)
    {
        var conn = cm.Get(req.ConnId!);
        if (conn is null)
            return new Response { Id = req.Id, Ok = false, Error = "conn_not_found" };
        var schema = req.Schema!.Replace("'", "''");
        var table  = req.Table!.Replace("'", "''");
        var sql =
            $"SELECT COLNAME, TYPENAME, DEFAULT, NULLS " +
            $"FROM SYSCAT.COLUMNS " +
            $"WHERE TABSCHEMA = '{schema}' AND TABNAME = '{table}' " +
            $"ORDER BY COLNO";
        return await RunSql(req.Id, conn, sql);
    }

    private static async Task<Response> RunSql(string reqId, DB2Connection conn, string sql)
    {
        var sw = Stopwatch.StartNew();
        try
        {
            using var cmd    = new DB2Command(sql, conn);
            using var reader = await cmd.ExecuteReaderAsync();

            var columns = Enumerable.Range(0, reader.FieldCount)
                .Select(i => reader.GetName(i))
                .ToList();

            var rows = new List<List<string?>>();
            while (await reader.ReadAsync())
            {
                var row = Enumerable.Range(0, reader.FieldCount)
                    .Select(i => reader.IsDBNull(i) ? null : reader.GetValue(i)?.ToString())
                    .ToList();
                rows.Add(row);
            }

            sw.Stop();
            return new Response
            {
                Id = reqId, Ok = true,
                Columns = columns, Rows = rows,
                ExecutionTimeMs = sw.ElapsedMilliseconds
            };
        }
        catch (Exception ex)
        {
            sw.Stop();
            return new Response
            {
                Id = reqId, Ok = false,
                Error = ex.Message,
                ExecutionTimeMs = sw.ElapsedMilliseconds
            };
        }
    }
}

File: db2-sidecar-java/src/main/java/com/aiterm/db2sidecar/CommandHandler.java
package com.aiterm.db2sidecar;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class CommandHandler {

    public static Response handle(Request req, ConnectionManager cm) throws Exception {
        switch (req.cmd) {
            case "connect":          return connect(req, cm);
            case "disconnect":       return disconnect(req, cm);
            case "ping":             return ping(req, cm);
            case "execute":          return execute(req, cm);
            case "list_schemas":     return listSchemas(req, cm);
            case "list_tables":      return listTables(req, cm);
            case "get_table_schema": return getTableSchema(req, cm);
            default:
                Response r = new Response();
                r.id = req.id;
                r.ok = false;
                r.error = "unknown_cmd:" + req.cmd;
                return r;
        }
    }

    private static Response connect(Request req, ConnectionManager cm) {
        String err = cm.connect(req.connId, req.connString, req.username, req.password);
        Response r = new Response();
        r.id = req.id;
        r.ok = (err == null);
        r.error = err;
        return r;
    }

    private static Response disconnect(Request req, ConnectionManager cm) {
        cm.disconnect(req.connId);
        Response r = new Response();
        r.id = req.id;
        r.ok = true;
        return r;
    }

    private static Response ping(Request req, ConnectionManager cm) {
        Connection conn = cm.get(req.connId);
        if (conn == null) {
            Response r = new Response();
            r.id = req.id;
            r.ok = false;
            r.error = "conn_not_found";
            return r;
        }
        try {
            try (Statement stmt = conn.createStatement()) {
                stmt.executeQuery("SELECT 1 FROM SYSIBM.SYSDUMMY1").close();
            }
            Response r = new Response();
            r.id = req.id;
            r.ok = true;
            return r;
        } catch (SQLException e) {
            Response r = new Response();
            r.id = req.id;
            r.ok = false;
            r.error = e.getMessage();
            return r;
        }
    }

    private static Response execute(Request req, ConnectionManager cm) {
        Connection conn = cm.get(req.connId);
        if (conn == null) {
            Response r = new Response();
            r.id = req.id;
            r.ok = false;
            r.error = "conn_not_found";
            return r;
        }
        return runSql(req.id, conn, req.sql);
    }

    private static Response listSchemas(Request req, ConnectionManager cm) {
        Connection conn = cm.get(req.connId);
        if (conn == null) {
            Response r = new Response();
            r.id = req.id;
            r.ok = false;
            r.error = "conn_not_found";
            return r;
        }
        String sql = "SELECT DISTINCT SCHEMANAME FROM SYSCAT.SCHEMATA " +
                     "WHERE DEFINERTYPE = 'U' ORDER BY SCHEMANAME";
        return runSql(req.id, conn, sql);
    }

    private static Response listTables(Request req, ConnectionManager cm) {
        Connection conn = cm.get(req.connId);
        if (conn == null) {
            Response r = new Response();
            r.id = req.id;
            r.ok = false;
            r.error = "conn_not_found";
            return r;
        }
        String schema = req.schema.replace("'", "''");
        String sql = "SELECT TABNAME, TYPE FROM SYSCAT.TABLES " +
                     "WHERE TABSCHEMA = '" + schema + "' ORDER BY TABNAME";
        return runSql(req.id, conn, sql);
    }

    private static Response getTableSchema(Request req, ConnectionManager cm) {
        Connection conn = cm.get(req.connId);
        if (conn == null) {
            Response r = new Response();
            r.id = req.id;
            r.ok = false;
            r.error = "conn_not_found";
            return r;
        }
        String schema = req.schema.replace("'", "''");
        String table  = req.table.replace("'", "''");
        String sql = "SELECT COLNAME, TYPENAME, DEFAULT, NULLS " +
                     "FROM SYSCAT.COLUMNS " +
                     "WHERE TABSCHEMA = '" + schema + "' AND TABNAME = '" + table + "' " +
                     "ORDER BY COLNO";
        return runSql(req.id, conn, sql);
    }

    private static Response runSql(String reqId, Connection conn, String sql) {
        long start = System.currentTimeMillis();
        try (Statement stmt = conn.createStatement()) {
            boolean hasResultSet = stmt.execute(sql);
            long elapsed = System.currentTimeMillis() - start;
            Response r = new Response();
            r.id = reqId;
            r.ok = true;
            r.executionTimeMs = elapsed;
            if (hasResultSet) {
                try (ResultSet rs = stmt.getResultSet()) {
                    ResultSetMetaData meta = rs.getMetaData();
                    int colCount = meta.getColumnCount();
                    List<String> columns = new ArrayList<>(colCount);
                    for (int i = 1; i <= colCount; i++) {
                        columns.add(meta.getColumnName(i));
                    }
                    List<List<String>> rows = new ArrayList<>();
                    while (rs.next()) {
                        List<String> row = new ArrayList<>(colCount);
                        for (int i = 1; i <= colCount; i++) {
                            String val = rs.getString(i);
                            row.add(rs.wasNull() ? null : val);
                        }
                        rows.add(row);
                    }
                    r.columns = columns;
                    r.rows = rows;
                }
            } else {
                r.affectedRows = (long) stmt.getUpdateCount();
            }
            return r;
        } catch (SQLException e) {
            long elapsed = System.currentTimeMillis() - start;
            Response r = new Response();
            r.id = reqId;
            r.ok = false;
            r.error = e.getMessage();
            r.executionTimeMs = elapsed;
            return r;
        }
    }
}
